pageserver/compaction/src/identify_levels.rs

```diff
@@ -0,0 +1,376 @@
+//! An LSM tree consists of multiple levels, each exponential larger than the
+//! previous level. And each level consists of be multiple "tiers". With tiered
+//! compaction, a level is compacted when it has accumulated more than N tiers,
+//! forming one tier on the next level.
+//!
+//! In the pageserver, we don't explicitly track the levels and tiers. Instead,
+//! we identify them by looking at the shapes of the layers. It's an easy task
+//! for a human, but it's not straightforward to come up with the exact
+//! rules. Especially if there are cases like interrupted, half-finished
+//! compactions, or highly skewed data distributions that have let us "skip"
+//! some levels. It's not critical to classify all cases correctly; at worst we
```

Comment:
There are no more half-finished compactions now that work of #5172, unless a compaction can be completed up to `RemoteTimelineClient::schedule_compaction_update` in some way half-way (did not at least pick up on this yet).

Cannot suggest because re-flowing :)

Reply:
Well, actually there is a case of producing some layers and then running into an error later: #4749.

I think fixing it will require keeping layers as "delete on drop tempfiles" (not the new `Layer`) until they have all been renamed as final paths (in two phases) and ready to be inserted into LayerMap.